### PR TITLE
[TACHYON-736] Add authenticated client user into a threadlocal variable

### PR DESCRIPTION
**Pull request by dongche**
JIRA: https://tachyon.atlassian.net/browse/TACHYON-736

After build the SASL transport in TACHYON-726(#1229), client could connect to server with its user info and server could verify the user based on authentication mechanism.
Then, if user is authenticated successfully, we should save the user and bind it with current thread for following RPC methods usage. The context is that, in thrift server, one client connection is allocated a thread during its lifecycle. This PR take care of this to assign the connected client user to a threadlocal.

Besides, for testing, a very simple thrift RPC service class is added (only one method). Not sure this is a good way. If it is fine, shall I need to upload the thrift definition file `xxx.thrift`?

@ooq , @apc999 , could you please help to review this PR? Thanks.

### CODE REVIEW

**Comment by ooq** [comment]
@dongche Yeah, so if you add Thrift generated java files, you should also add the .thrift files. But I don't think you want to add all those files just for this test... just piggyback on the original thrift file?

File: common/src/main/java/tachyon/security/PlainSaslServer.java [hunk 1]
```diff
@@ -182,6 +191,7 @@ public void handle(Callback[] callbacks) throws IOException, UnsupportedCallback
 
       if (ac != null) {
         ac.setAuthorized(true);
+        RemoteClientUser.set(username);
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: 1, I think `RemoteClientUser` should be set in the `PlainSaslServer#evaluateResponse` as opposed to here. 2, when the authorization ID and authentication ID are different, which one should you pick for the remote client username?

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 2]
```diff
@@ -135,8 +140,20 @@ public void simpleAuthenticationTest() throws Exception {
     client.open();
     Assert.assertTrue(client.isOpen());
 
+    // check user 'anyone'
+    verifyClientUser(client, "anyone");
+
+    // another connection by user 'anyone1'
+    TTransport client1 = createClient(mTachyonConf, "anyone1", "whatever");
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: I don't actually see the reason to check another username here, does it actually add more code coverage?

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java [hunk 3]
```diff
@@ -231,8 +251,20 @@ public void customAuthenticationExactNamePasswordMatchTest() throws Exception {
     client.open();
     Assert.assertTrue(client.isOpen());
 
+    // check user 'tachyon'
+    verifyClientUser(client, "tachyon");
+
+    // another connection by user 'tachyon1'
+    TTransport client1 = createClient(mTachyonConf, "tachyon1", "correct-password1");
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Similar as above.

**Comment by ooq** [comment]
@dongche This PR looks good, and I like that we have separated RemoteClientUser and LoginUser! What you said about the thrift interface is a valid concern, and I feel you should piggyback on the original thrift service, either through adding another RPC or use the old RPC to do an integration test.

**Comment by dongche** [comment]
@ooq Sure, using the original thrift file is convenient. Only one concern was that when we select one method of `MasterService.Iface` to be overridden for test usage, the method name may be not meaningful. 

Maybe it is not a problem for just test purpose. So let's leverage the original thrift file like you suggest. How about re-implements this method `public String user_getUfsAddress()` and let it return the username? This method has the same contract as the `whoAmI` method in the patch.

File: common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java (hunk 2 above)
Comment (dongche) [inline common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java]: Not sure adding more coverage... Seems not.

Checking another username here was intent to ensure server could handle multiple client connection at the same time and maintain corresponding user info correctly.

**Comment by ooq** [comment]
What do you mean re-implement? I would not modify the behavior of original RPC interface.. You know what, I think you should create a JIRA for this test, and let's test it when we have some authorization. (Authorization tests will naturally cover this test)

**Comment by dongche** [comment]
@ooq Yes, authorization tests will cover this. But it would be better if we have some test here to cover the added code. In the updated patch, I tried another way to realize it (by mocking a MasterService - the original PRC interface). How does that sound?

One bad smell is that the method name and its behavior does not match. Looks a litter wired...
If this solution is not preferred, let's postpone to test it when we have some authorization.

BTW, other comments are also address in the updated patch. Thanks for your review!

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 1 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: 1, Done. 2, Good question. Thanks! We should pick authorization ID. Updated in latest patch.

**Comment by ooq** [comment]
Jenkins, test this please.

**Comment by ooq** [comment]
@dongche Good hack! I left more comments.

**Comment by ooq** [comment]
@dongche I see you cannot easily extend MasterServerHandler. Still, I think having a method `user_getUfsAddress` but actually return the remote user name is a bad idea. Nobody will understand this code later on.

**Comment by dongche** [comment]
@ooq Yes, inconsistency of method `user_getUfsAddress`'s name and behavior is bad smell in code. 

Not sure if there is other way to test this. So I will withdraw the test part in this patch and only keep the code about get/set `RemoteClientUser`. Testing it will be done in follow-up Jira after we have some authorization code.
Is this ok?

**Comment by ooq** [comment]
@dongche Sounds good!

**Comment by dongche** [comment]
Patch updated. (1, withdraw test. 2, address comment about checking case of authzid equals authcid)

File: common/src/main/java/tachyon/security/PlainSaslServer.java [hunk 4]
```diff
@@ -39,7 +38,7 @@
  * 3.Write a JCA provider that registers the factory
  */
 public class PlainSaslServer implements SaslServer {
-  private String mAuthcid;
+  private String mAuthzid;
```
Comment (ooq) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: I think the username is authorization ID, which is mAuthcid. Why do we want to change it here? @shenguoquan Can you also review this? Thanks!

**Comment by ooq** [comment]
@dongche Thanks for the update, just left one comment. Also can you create a JIRA for adding the tests later on?

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 4 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: I agree username should be authorization ID.  But it is not `mAuthcid`.

Actually, `mAuthcid` is authentication ID, and `mAuthzid` is authorization ID. Original code logic save the authentication ID(`mAuthcid`) as username, which is not expected beharior. So it needs to be changed here.

**Comment by dongche** [comment]
@ooq Thanks for continuous review! Add a reply to the comment.

With regards to create a JIRA for test, I was thinking either creating it in the simple-authentication umbrella JIRA TACHYON-593, or creating it in the upcoming umbrella JIRA about authorization first stage talked in the sync up meeting (target to the authorization branch)? I guess the latter one?

File: common/src/main/java/tachyon/security/RemoteClientUser.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ * from client side and take back it when connection is closed. During the thread alive cycle,
+ * all the RPC happens in this thread. These RPC methods implemented in server side could
+ * get the client user by this class.
+ */
+public class RemoteClientUser {
+
+  /**
+   * A ThreadLocal variable to maintain the client user along with a specific thread.
+   */
+  private static final ThreadLocal<User> USER_THREAD_LOCAL = new ThreadLocal<User>();
```
Comment (apc999) [inline common/src/main/java/tachyon/security/RemoteClientUser.java]: Naming: this field is static final and currently named after the constant name convention (`SOME_NAME`). However this filed particularly is not really a const as its state changes. Maybe better to have it as `mUserThreadLocal`.

File: common/src/main/java/tachyon/security/RemoteClientUser.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * An instance of this class represents a client user connecting to Tachyon service.
+ *
+ * It is maintained in a ThreadLocal variable based on the Thrift RPC mechanism.
+ * {@link org.apache.thrift.server.TThreadPoolServer} allocates a thread to serve a connection
+ * from client side and take back it when connection is closed. During the thread alive cycle,
+ * all the RPC happens in this thread. These RPC methods implemented in server side could
+ * get the client user by this class.
+ */
+public class RemoteClientUser {
```
Comment (apc999) [inline common/src/main/java/tachyon/security/RemoteClientUser.java]: Question: will we have `LocalClientUser` in the future? If not, I prefer to name this class after something like `ClientUser` for simplicity. If there is `LocalClientUser` coming, discard my comment here :)

File: common/src/main/java/tachyon/security/PlainSaslServer.java [hunk 7]
```diff
@@ -76,29 +75,36 @@ public String getMechanismName() {
...
 
       Callback[] cbList = {nameCallback, passwordCallback, authCallback};
       mHandler.handle(cbList);
       if (!authCallback.isAuthorized()) {
         throw new SaslException("AuthorizeCallback authorized failure");
       }
+      mAuthzid = authCallback.getAuthorizedID();
+
+      // After verification succeeds, a user with this authz id will be set to a Threadlocal.
+      RemoteClientUser.set(mAuthzid);
```
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: Question about using this Threadlocal storage: it seems we only `set` and `remove`, where do we really use this thread local variable?

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 4 above)
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: Please do javadoc this member variable `mAuthzid`, as I didn't realize the difference that it is for authorization ID not for authentication ID from its name.

Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: Also, does that mean this variable is not currently used but for future authorization usage? If that is the case, please also javadoc that.

File: common/src/main/java/tachyon/security/PlainSaslServer.java [hunk 8]
```diff
@@ -76,29 +75,36 @@ public String getMechanismName() {
...
-      if (mAuthcid == null || mAuthcid.isEmpty()) {
+      if (authcid == null || authcid.isEmpty()) {
         throw new IllegalStateException("No authentication identity provided");
       }
       if (passwd == null || passwd.isEmpty()) {
         throw new IllegalStateException("No password provided");
       }
       if (authzid == null || authzid.isEmpty()) { // authzid = authcid
-        authzid = mAuthcid;
+        authzid = authcid;
```
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: So currently we always assign authcid to authzid, right?
We should document this some where (maybe in the class description)?

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 7 above)
Comment (apc999) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: Another dumb question here: where is `PlainSaslServer` used? e.g., is it run with the client thread, or the master thread? I guess I am trying to figure out this thread-local storage is local to which thread.

**Comment by apc999** [comment]
@dongche , Thanks for the patch. I left some readability comments. Mostly, I try to understand this patch better with some questions. Thanks!

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 4 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: Done.

File: common/src/main/java/tachyon/security/RemoteClientUser.java (hunk 5 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/RemoteClientUser.java]: Done

File: common/src/main/java/tachyon/security/RemoteClientUser.java (hunk 6 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/RemoteClientUser.java]: Make sense. Thanks. Renamed to `AuthorizedClientUser`. Currently no plan for `LocalClientUser`

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 7 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: This thread local variable will be used by RPC methods (such as file operation methods) after connection is built. e.g. The user could be retrieved, and then associated with the created file or checked whether this user could operate on a file.

`PlainSaslServer` is used in the client thread and its method `evaluateResponse` runs in client thread, so the thread local storage is local to the connected client thread.

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 8 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: Right. Doc-ed.

**Comment by dongche** [comment]
@apc999 , thanks for your review and feedback! Patch updated to add javadoc and rename variables to make the code more readable. 

Add some reply to the comments. Please let me know if any confusion, thanks.

File: common/src/main/java/tachyon/security/PlainSaslServer.java (hunk 7 above)
Comment (dongche) [inline common/src/main/java/tachyon/security/PlainSaslServer.java]: When client try to open the connection, a thread in server is allocated, in which `PlainSaslServer`'s `evaluateResponse` authenticate the user - 
If fails, connection fails, thread is take back.
If succeed, connection is built, thread will be alive util client close connection.

**Comment by apc999** [comment]
+1 LGTM.
Thanks for the update!

**Comment by ooq** [comment]
+1 LGTM. Thanks for the patch!

**Comment by ooq** [comment]
@haoyuan Merge this?

**Pull request merged**